_posts/2015-03-23-design-patterns-part-2.markdown

```diff
@@ -0,0 +1,216 @@
+dependencies passed on initialization (but they'll still be faster than if we booted up the entire app).
+
+Our tests for the factory are pretty basic.  Since the factory only creates new instances of things, we just need to verify that it’s working properly.  Something like this should be sufficient for a unit test.
+
+{% highlight ruby %}
+
+describe UserFactory do
+
+  it 'constructs a create_user context' d
+      expect(
+     described_class.create_user
+     ).to be_a CreateUser
+  end
+
+end
+
+{% endhighlight %}
+
+Some developers don’t even think it’s worth writing test for factories, since tests are meant to verify behavior and factories have no behavior.  It’s really up to you as to what you feel is best for your app.
+
+At Sittercity, we write factory classes that are responsible for instantiating our contexts.  We adopt the convention of keeping these factories as modules rather than as full-fledged classes.
+
+In order to make our test pass, we would end up writing the following code for our factory.
+
+{% highlight ruby %}
+
+require ‘create_user’
+require ‘mailer’
+require ‘AR_user_repository’
+
+module UserFactory
+
+   self.create_user
+    CreateUser.new(ARUserRepository.new, UserMailer.new)
+   end
+
+end
+
+{% endhighlight %}
+
+The only thing the factory does is to make it easy to get an instance of our context.   
+
+Back in our controller, we'll have something like this:
+
+{% highlight ruby %}
+
+class UsersController
+  def create
+    context = UserFactory.create_user
+    user = context.call(params[:first_name], params[:last_name], params[:email])
+  end
+end
+
+{% endhighlight %}
+
+Despite the fact that creating a user involves a context class, our repository, our entity, the User model, and the mailer class, our controller
+only need to know two things.  First, it needs to know that the UserFactory has a class method named "create\_user" and second, it needs to know
+that the context for creating user has one public method named “call,” which accepts three arguments (last\_name, first\_name, and email).  The controller can be blissfully unaware of everything beyond those two things.
+
+## Conclusion
+
+	Our use of the command pattern plays an integral role in decoupling our software of the web framework.  We no longer rely on business logic that lives inside Rails models.  Instead, we have flexible and robust contexts that describe what our app does.  These are not bound to any particular model.  Instead, they are constructed with any dependencies they might have. Using dependency injection allows for this, while also providing an opportunity to write screaming fast tests.  Finally, we pull it all together with a factory that any class can easily call.  We’re laying down a foundation that will give us back control of our software and make it resilient to the changes that await it.
```

Comment:
This needs to be fixed...try looking at it in the browser, this formatting puts it all into a big code block looking thing.